crates/sui-core/src/authority_server.rs

```diff
@@ -622,6 +616,160 @@ impl Validator for ValidatorService {
             .get_cache_reader()
             .get_sui_system_state_object_unsafe()?;
 
-        return Ok(tonic::Response::new(response));
+        Ok(tonic::Response::new(response))
+    }
+
+    async fn handle_traffic_req(
+        &self,
+        connection_ip: Option<SocketAddr>,
+        proxy_ip: Option<SocketAddr>,
+    ) -> Result<(), tonic::Status> {
+        if !self.traffic_controller.check(connection_ip, proxy_ip).await {
+            // Entity in blocklist
+            Err(tonic::Status::from_error(SuiError::TooManyRequests.into()))
+        } else {
+            Ok(())
+        }
+    }
+
+    fn handle_traffic_resp<T>(
+        &self,
+        connection_ip: Option<SocketAddr>,
+        proxy_ip: Option<SocketAddr>,
+        response: &Result<tonic::Response<T>, tonic::Status>,
+    ) {
+        let result: SuiResult = if let Err(status) = response {
+            Err(SuiError::from(status.clone()))
+        } else {
+            Ok(())
+        };
+
+        self.traffic_controller.tally(TrafficTally {
+            connection_ip,
+            proxy_ip,
+            result,
+            timestamp: Utc::now(),
+        });
+    }
+}
+
+/// Implements generic pre- and post-processing. Since this is on the critical
+/// path, any heavy lifting should be done in a separate non-blocking task
+/// unless it is necessary to override the return value.
+#[macro_export]
+macro_rules! handle_with_decoration {
+    ($self:ident, $func_name:ident, $request:ident) => {{
+        // extract IP info. Note that in addition to extracting the client IP from
+        // the request header, we also get the remote address in case we need to
+        // throttle a fullnode, or an end user is running a local quorum driver.
+        let connection_ip: Option<SocketAddr> = $request.remote_addr();
+
+        // This should never happen except perhaps in simtest or some other non-standard
+        // environment. If we are seeing this, we should investigate
+        // TODO: add metric here
+        if connection_ip.is_none() {
+            if cfg!(all(test, not(msim))) {
+                panic!("Failed to get remote address from request");
+            } else {
+                error!("Failed to get remote address from request");
+            }
+        }
+
+        let proxy_ip: Option<SocketAddr> =
+            if let Some(op) = $request.metadata().get("x-forwarded-for") {
+                match op.to_str() {
+                    Ok(ip) => match ip.parse() {
+                        Ok(ret) => Some(ret),
+                        Err(e) => {
+                            error!("Failed to parse x-forwarded-for header value to SocketAddr: {:?}", e);
+                            None
```

Comment:
i think this should return an error, there's no reason a client would accidentally insert a malformed IP